### ISSUE TITLE
[BUG] Print fails using raft supports

### ISSUE THREAD

**Issue by rado79** [jschuh/klipper-macros#48]
**Describe the bug**
I slice my models with Cura 5.1
Using supports with brim or skirt works a s aspected.
If i choose Raft (in my Case 4 Layers) i get this error and print fails.

**Klipper error messages**
00:30
SDCARD_RESET_FILE cannot be run from the sdcard
00:30
Error on 'SET_PRINT_STATS_INFO CURRENT_LAYER="-4"': CURRENT_LAYER must have minimum of 0
00:30
Error on 'SET_PRINT_STATS_INFO CURRENT_LAYER="-4"': CURRENT_LAYER must have minimum of 0

**Klipper config**
[config (2).zip](https://github.com/jschuh/klipper-macros/files/10787927/config.2.zip)
[CFFFP_mandibula.zip](https://github.com/jschuh/klipper-macros/files/10787982/CFFFP_mandibula.zip)

**Comment by jschuh** [comment]
Thanks, this is useful. People have filed bugs about Cura passing negative layer heights, but I never knew why (I don't use Cura) and no one seemed to know what settings were the cause. Now that I know it's because Cura uses negative layer values for rafts, I can think of a better way to handle it.

**Comment by jschuh** [comment]
Please let me know if bfe55c5 fixes the issue.

**Comment by rado79** [comment]
No, same issue again.

Do we need this semicolon?
![semicolon](https://user-images.githubusercontent.com/51396577/220667533-75e8c209-362d-4a95-8fe9-472ca4fef71d.png)

**Comment by jschuh** [comment]
Even if it's not fixed you shouldn't be getting exactly the same error. Could you paste in the exact error text? Also, sorry if this is a dumb question, but I just need to verify you did restart Klipper after updating klipper-macros and attempting the print again?

And yes, that semicolon is required. The gcode you attached shows that the substitutions are all working correctly. The problem is just that Cura uses negative layer values for the raft.

**Comment by rado79** [comment]
![image](https://user-images.githubusercontent.com/51396577/220673307-173b66b4-7e00-4524-8591-e9ac2c938b1e.png)

**Comment by rado79** [comment]
This is while purge line prints
![image](https://user-images.githubusercontent.com/51396577/220675256-0bba8482-4aa6-488f-aaaf-5edcc1064329.png)
It shows Layer 2 of 774

**Comment by rado79** [comment]
After purge line, the print stops with the error
![image](https://user-images.githubusercontent.com/51396577/220676218-2f06d71a-2a00-4ff9-a424-29823b6dbc59.png)
shows Layer 112 of 774

**Comment by rado79** [comment]
And i restartet everything again one more time, same error

**Comment by rado79** [comment]
I dont know if my update process i the right thin i do. I copy the text from layers.cfg of your commit,
and paste it in my Klipper config to layers.cfg

**Comment by rado79** [comment]
[layers.zip](https://github.com/jschuh/klipper-macros/files/10805783/layers.zip)

**Comment by rado79** [comment]
> Oh, could you just do a `git pull` to update so I can be sure it's running the right version? You may have to do a `git reset --hard HEAD` first to revert any local changes to the macros.

if i do so, do i lose changes like "G1 E12 F1000" direct before purgeline begins?

**Comment by jschuh** [comment]
Yes, it would revert your local changes. The thing is that I can't really figure out what's going on if the macros aren't a stock installation of the most current version.

**Comment by rado79** [comment]
okay, i wil do it like you said before

**Comment by rado79** [comment]
![image](https://user-images.githubusercontent.com/51396577/220686677-edd26357-debd-4ca8-a8d1-c1b6d99ae790.png)

**Comment by jschuh** [comment]
I just landed 54b137d, which simply ignores negative layer values. The previous change tried to do some extra validation, but I decided that just isn't worth the trouble.

**Comment by jschuh** [comment]
I'd been thinking about adding a config option for user-supplied gcode immediately before the purge/print starts, since I noticed that's a common customization. I'll try to test and land that today.

**Comment by rado79** [comment]
After a hard reset and update to the latest commit, Raft prints as expected.
Layer counter starts with first layer of the model.
Well done!

**Comment by jschuh** [comment]
Ah, interesting. I'd thought about adding purge priming like Prusa printers do, but I hadn't needed it on any of my printers. It's an easy change to incorporate into `DRAW_PURGE_LINE`, so I'll look into it.

That stated, I still think generic user-defined gcode before the print is independently needed for things like docking probes and the ERCF.
